Add an interface for surfacing tool calls

cpsievert opened the pull request:
Addresses https://github.com/posit-dev/chatlas/issues/33
Related https://github.com/posit-dev/shinychat/issues/31

This PR adds a few things to help with surfacing tool requests and results in response content.

1. A new `ToolResult()` class, which allows for: 
    * Control over how results get formatted when sent to the model 
    * Yielding of additional content for the user (i.e., the downstream consumer of a `.stream()` or `.chat()`) to display when the tool is called.
2. A new `on_request` parameter for `.register_tool()`. When tool is requested, this callback executes, and the result is yielded to the user.
3. A new `Chat.on_tool_request()` method for registering a default tool request handler.

Here is a basic Shiny example:

```python
import asyncio
from shiny.express import ui
from chatlas import ChatOpenAI, ToolResult

chat_model = ChatOpenAI()

async def get_current_temperature(latitude: float, longitude: float):
    # Simulate a slow operation
    await asyncio.sleep(4)

    return ToolResult(
        "72°F.",
        user="""
<details class="pb-3">
<summary>Tool result</summary>
The <code>get_current_temperature()</code> tool returned: 72°F.
</details>\n\n\n\n
""",
    )

chat_model.register_tool(
    get_current_temperature,
    on_request=lambda request: "Requesting the current temperature...\n\n",
)

chat = ui.Chat(id="chat", messages=["**Hello!** How can I help you today?"])
chat.ui()
chat.update_user_input(value="How is the weather in Duluth, MN today?")

@chat.on_user_submit
async def handle_user_input(user_input: str):
    response = await chat_model.stream_async(user_input)
    await chat.append_message_stream(response)
```



https://github.com/user-attachments/assets/8e630c66-302a-4a08-8d87-b788a8c6fb55





### TODO

- [ ] Do we care about implications for MCP?
- [ ] Documentations and examples
- [x] Tests

cpsievert committed:
First pass at an interface for surfacing tool calls

cpsievert commented on chatlas/_content.py:
```diff
@@ -224,9 +232,9 @@ def __repr__(self, indent: int = 0):
     def get_final_value(self) -> str:
-        value, _language = self._get_value_and_language()
-        return value
+        return self._get_value()
```
This is an intentional, and somewhat orthogonal, change in behavior that was introduced in #30 where we started sending a `pformat()` version if the value can be loaded as JSON. Now, instead, we always send the `str()`'d result for consistency sake (as we did before #30). I _think_ @schloerke's intention in #30 was to just get better formatting of tool calls when `echo="all"`, but I might be wrong (sorry for not catching this earlier).

cpsievert replied:
Moving to #70

gadenbuie commented:
I like the idea of a `ToolResult` class, but I'm definitely conflicted about having to compose `ToolResult` _inside_ the actual tool function, but the simplicity of the approach is convincing.

For completeness and generality, I think `ToolResult` should have three key properties:

* `.data`: The actual data from the tool call. E.g. if you called a weather API, this would be the dictionary with the API response.
* `.result` or `.text` (unsure of the name): The text returned to the LLM. We could also use a `__str__` method that looks for this first or calls `str(self.data)`.
* `.html`: The HTML display output, which could be raw HTML or `htmltools.Tag`s, etc.

cpsievert committed:
Add serialized_value to ToolResult(); tweak naming; refactor

cpsievert commented:
> .data: The actual data from the tool call. E.g. if you called a weather API, this would be the dictionary with the API response.

This previously was `assistant` cause I was originally thinking I would name the yielded value `user`. Just changed it to `value`.

> .result or .text (unsure of the name): The text returned to the LLM. We could also use a __str__ method that looks for this first or calls str(self.data).

Added a `.serialized_value` that captures this idea.

> .html: The HTML display output, which could be raw HTML or htmltools.Tags, etc.

I'm currently calling this `.response_output`. `.html` doesn't feel quite right since `chatlas` doesn't care how the yielded value gets rendered (it would be totally valid for the assistant and tools to be emitting `json`, `rtf`, `xml`, or whatever)

cpsievert committed:
Quote type

cpsievert committed:
Update tests

gadenbuie commented:
> I'm currently calling this `.response_output`. `.html` doesn't feel quite right since `chatlas` doesn't care how the yielded value gets rendered

I mean, this is the crux of why it feels a little weird to be putting `ToolResult` in the tool call function.

We might be misunderstanding each other, but I don't think the names I'm proposing match the renaming your proposing. I'll try rephrasing, but I've also refined my naming proposal down to `.data`, `.assistant` 

1. The `.data` attribute is the result of the function called in the tool, as in the raw Python object, such as the dictionary returned by calling an API or the numerical result of a computation.
2. The `.assistant` (what I called `.text`) is the text sent to the LLM. I don't think `.serialized_value` is quite right here, because it doesn't have to be `.data` just serialized to text. There could be other transformations that happen to `.data` to turn it into the text seen by the LLM. The idea is that you're explicitly transforming the text for consumption by the LLM. (But we _can_ just by default serialize `.data`.)
3. Finally, `.display` (or what I called `.html`) is the text or HTML shown to a user. This could be HTML for Shiny or it could be pretty text for a terminal UI.
   * That said, I do think it's better to have `.html`, `.text` and other output-specific names because the content itself will inherently be useful in specific contexts. In other words, if you wanted to have different output for Shiny and for a terminal UI, you'd wouldn't want to put that all in `.display`.
   * Maybe `.display` is itself a class with sub-properties for this, e.g. `.display.html` for Shiny and `.display.text` for user-facing text.

cpsievert commented:
> Finally, .display (or what I called .html) is the text or HTML shown to a user. 

With that naming, it'd be reasonable to think that it displayed in the console, but it doesn't. It only gets included in the _return value_ of `.chat()` and `.stream()`. The consumer of that return value determines how to (or not to) display it.

cpsievert committed:
Update ChatResponse/ChatResponseAsync to be iterable of Stringable (instead of str)

cpsievert committed:
Update .stream() and .stream_async() to return generators not ChatResponse

I don't it was necessary in the first place, and leads to inefficient use of memory

cpsievert committed:
Add unit tests

gadenbuie commented:
> > Finally, .display (or what I called .html) is the text or HTML shown to a user.
> 
> With that naming, it'd be reasonable to think that it displayed in the console, but it doesn't. It only gets included in the _return value_ of `.chat()` and `.stream()`. The consumer of that return value determines how to (or not to) display it.

But I'm saying that it _could_ be shown in the console? Okay, let me back up. The point is that calling a tool gives you 

1. A literal object in Python

and then as the tool owner you get to make choices about:

1. What is shown to the LLM
2. What is shown to the user

The LLM needs text but that text might be different from a textual display to the user. In fact, for the user-facing display, there might be several different venues with different formatting. The two we've thought of so far are "text in the console" and "html in the browser".

cpsievert commented:
> But I'm saying that it could be shown in the console? 

The current way to display tool calls is `echo="all"`, and if you note [this comment](https://github.com/posit-dev/chatlas/pull/69/files#diff-06d808cea8365cc6d2087c3085f7a8e74d296eb80048950dcc21c99b8e3dccccR95-R98), I'm pointing out that we _could_ allow for customizing display to console, but that's not the priority here.

Anyway, I don't disagree that customizing display would be a nice to have, but even in that case, it's important to distinguish what gets `echo`ed and what gets returned since they're two different targets.

gadenbuie commented:
Yeah, I agree that we shouldn't prioritize customizing `echo="all"`. I actually don't think we should allow that at all; it should show the output the LLM saw.

I think ultimately we're saying the same thing, or at least wanting `ToolResult` to include the same pieces. I think `.value`, `.response_output` and `.serialized_value` are confusing names that don't clearly communicate intent or how the values are used. They each could be confused for at least one of the others. I'm just trying to nudge the names in a more user-friendly direction. Sorry if I've been over-explaining

cpsievert committed:
Smarter formatting default for tool results, better way to customize/opt-out, and better naming/docs

cpsievert committed:
Merge branch 'main' into surface-tool-calls

cpsievert committed:
Update changelog

wch commented:
Oh, and one more thing about being able to pass in arbitrary objects to the tool. Suppose you want to use the same tool in a Shiny app, and in a console app.

Let's say the tool looks like this, where it takes an `emit` argument which is a function:

```py
# Define the tool in a scope outside of the app's server code
def run_query(query: str, data: pd.DataFrame, emit: Callable[[str], Awaitable[None]]):
    emit(f"Starting query: f{query}...")
    ...
    emit("Running query...")
    ...
    emit("Finished query...")
    return ...
```  

The `emit` function needs to be provided to the tool, and it can take different forms.

In a console app, you might just pass in the `chat_model.emit` method:

```py
df = pd.DataFrame(...)

chat_model = ChatAnthropic(...)

@chat_model.register_tool(run_query, extra_args = {
    "data": dynamic(lambda: df)
    "emit": chat_model.emit
})
```

But in Shiny, you might do something fancier with those messages. In this case, it might wrap them in `<code>` tags, and send it to the Shiny chat UI stream, instead of going through the chatlas message stream:

```py
def server(input, output):
    # Suppose the value of df can change over time
    df = pd.DataFrame(...)

    # This is the Shiny chat object
    chat = ui.Chat(...)

    async def append_code_to_chat(txt: str):
        await chat.append_message_stream(ui.tags.code(txt))


    # The chatlas chat object
    chat_model = ChatAnthropic(...)

    @chat_model.register_tool(run_query, extra_args = {
        "data": dynamic(lambda: df)
        "emit": append_code_to_chat
    })
```


This would also the tool caller to define their own functions for many purposes, like displaying progress, or getting user input:

If you have a long computation that needs to display progress, then that progress could be implemented one way at the console, another way in Shiny, and yet another in Streamlit:

```py
def long_computation(x: int, progress: Callable[[int], Awaitable[None]]):
    progress(0)
    ...
    progress(33)
    ...
    progress(66)
    ...
    progress(100)
    return ...
```

Or say the tool needs to get user input to confirm something

```py
def ask_user_yes_no(
    msg: str,
    confirm: Callable[[str], Awaitable[bool]]
):
    user_response = await confirm(msg)
    return user_response
```

*****

In a discussion with @jcheng about this, he pointed out that we can already do some of these things, with currying:

```
# The tool function, defined outside of the Shiny app
async def run_query(
    query: str,
    data: pd.DataFrame,
    emit = Callable[[str], Awaitable[None]]
):
    """Runs a SQL query on data"""
    await emit(f"Starting query: f{query}...")
    ...
    await emit("Running query...")
    ...
    await emit("Finished query...")
    return ...

# Function for wrapping data and emit
def make_run_query(
    data: Callable[[], pd.DataFrame],
    emit: Callable[[str], Awaitable[None]]
):
    def wrapped_run_query(query):
        run_query(query, data(), emit)

    return wrapped_run_query
    

## Using in Shiny
def server():
    async def append_code_to_chat(txt: str):
        await chat.append_message_stream(ui.tags.code(txt))

    chat_model.register_tool(make_run_query(df, append_code_to_chat))


## Using in the terminal
chat_model.register_tool(make_run_query(df, print)) # Prints to console
```

Both of the uses above, in Shiny and the terminal, output directly to their respective UIs. But if we want to modify the chatlas object's output stream, it might be something like this:


```py
## Emit to chat_model's output stream, at the chatlas level
chat_model.register_tool(make_run_query(df, chat_model.emit))
```

*****

And finally, one other possibility that we discussed, where if the tool function takes a parameter named `_chat` of type `Chat`, we automatically pass in the chatlas `Chat` object. So defining and registering the tool in this case is very simple:

```py
# This version of run_query will emit to the chatlas object's output stream
async def run_query(query: str, _chat: Chat):
    """Runs a SQL query on data"""
    await _chat.emit(f"Starting query: f{query}...")
    ...
    await _chat.emit("Running query...")
    ...
    await _chat.emit("Finished query...")
    return ...


chat_model.register_tool(run_query)
```
